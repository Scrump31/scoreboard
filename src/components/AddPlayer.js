import React, { Component } from 'react';
import _ from 'lodash';
import { Col } from 'react-bootstrap';
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';
import { Field, reduxForm } from 'redux-form';
import { reset } from 'redux-form';
import { addNewPlayer } from '../actions';

class AddPlayer extends Component {

  renderPlayerNameField(field) {
    const { meta: { invalid, error, touched } } = field;
    return (
      <div className="form-group">
        <div className="input-group">
          <input
            type="text"
            className="form-control"
            placeholder="Add a new player..."
            {...field.input}
          />
          <span className="input-group-btn">
            <button
              className="btn btn-default"
              type="submit"
            >Add
          </button>
          </span>
        </div>
        <div className="playerValidate">
          { touched && invalid ? error : '' }
        </div>
      </div>
    );
  }

  createNewPlayer(values) {
    const { dispatch } = this.props;
    this.props.addNewPlayer(
      {
        name: values.playerName,
        score: 0,
        id: _.uniqueId(),
      },
    );
    return dispatch(reset('AddPlayerForm'));
  }

  render() {
    const { handleSubmit } = this.props;
    return (
      <Col md={12}>
        <form onSubmit={handleSubmit(this.createNewPlayer.bind(this))}>
          <Field
            name="playerName"
            component={this.renderPlayerNameField}
          />
        </form>
      </Col>
    );
  }
}

function mapDispatchToProps(dispatch) {
  return bindActionCreators({ addNewPlayer }, dispatch);
}

function validate(values) {
  const errors = {};

  if (!values.playerName) {
    errors.playerName = 'Please Enter A Player Name';
  }

  return errors;
}

export default reduxForm({
  validate,
  form: 'AddPlayerForm',
})(
  connect(null, mapDispatchToProps)(AddPlayer),
);
