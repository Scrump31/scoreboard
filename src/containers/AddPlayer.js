import React, { Component } from 'react';
import { FormGroup, FormControl, InputGroup, Button } from 'react-bootstrap';

import _ from 'lodash';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { addNewPlayer } from '../actions';

export class AddPlayer extends Component {
  state = {
    value: '',
  };

  handleChange = e => this.setState({ value: e.target.value });

  generatePlayer = () => {
    this.props.createPlayer({
      id: _.uniqueId(),
      name: this.state.value,
      score: 0,
    });
    this.setState({ value: '' });
  };

  handleSubmit = e => {
    e.preventDefault();
    if (this.state.value) {
      this.generatePlayer();
    } else {
      return null;
    }
  };
  render() {
    return (
      <form onSubmit={this.handleSubmit}>
        <FormGroup>
          <InputGroup>
            <FormControl
              type="text"
              value={this.state.value}
              placeholder="Add a new player..."
              onChange={this.handleChange}
            />
            <InputGroup.Button>
              <Button type="submit">Add</Button>
            </InputGroup.Button>
          </InputGroup>
        </FormGroup>
      </form>
    );
  }
}

const mapDispatchToProps = dispatch => ({
  createPlayer: player => dispatch(addNewPlayer(player)),
});

AddPlayer.propTypes = {
  createPlayer: PropTypes.func.isRequired,
};

export default connect(
  null,
  mapDispatchToProps,
)(AddPlayer);
