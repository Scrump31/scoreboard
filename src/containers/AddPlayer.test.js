import React from 'react';
import { mount } from 'enzyme';
import { AddPlayer } from './AddPlayer';

describe('<AddPlayer />', () => {
  const mockAddNewPlayer = jest.fn();

  it('calls mockAddNewPlayer() when name is entered and "Add" button clicked', () => {
    const component = mount(<AddPlayer createPlayer={mockAddNewPlayer} />);
    component
      .find('input')
      .simulate('change', { target: { value: 'new player' } });
    component.find('form').simulate('submit', { preventDefault() {} });
    expect(mockAddNewPlayer).toHaveBeenCalled();
    component.unmount();
  });
});
