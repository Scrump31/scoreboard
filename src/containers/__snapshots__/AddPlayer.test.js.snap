// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AddPlayer /> renders without crashing 1`] = `
<form
  onSubmit={[Function]}
>
  <FormGroup
    bsClass="form-group"
  >
    <InputGroup
      bsClass="input-group"
    >
      <FormControl
        bsClass="form-control"
        componentClass="input"
        onChange={[Function]}
        placeholder="Add a new player..."
        type="text"
        value=""
      />
      <InputGroupButton
        bsClass="input-group-btn"
      >
        <Button
          active={false}
          block={false}
          bsClass="btn"
          bsStyle="default"
          disabled={false}
          type="submit"
        >
          Add
        </Button>
      </InputGroupButton>
    </InputGroup>
  </FormGroup>
</form>
`;
